### DISCUSSION TITLE
Mesh merge on upbge.

### DISCUSSION THREAD

**Discussion by HolandonHdunto** [UPBGE/upbge#1693] [Ideas/Features for UPBGE 0.2.5+]
I don't know if upbge has this functionality, but the possibility of mesh merging would be interesting, which merges the mesh of a given object with another and excludes the original object within the bge. It is very useful to improve game performance by reducing DRAW CALLS as each individual object requires calculation of vertices, uv maps, local position, global position, materials and etc, and with mesh merge you reduce the number of calculations. And maybe if you have a handy way of isolating the vertices of a mesh by storing the index and an array, you can remove or add a mesh set.

Alias ​​would be great a new option of fixed objects in addition to the static ones, the fixed ones would automatically be joined with other fixed objects in a super mesh. If the fixed object is added to a group, as soon as the game starts, it is joined with the others.

**Comment by HolandonHdunto** [comment]
https://upbge.org/#/documentation/docs/v0.2.5/api/bge.types.KX_BatchGroup.html#KX_BatchGroup

but there is no way to get the resulting mesh.

**Comment by BluePrintRandom** [comment]
so there is a optimization that all objects materials are baked to a single atlas, so all materials are 1.

this makes it so batching matters, (batching items without the same material still results in multiple draw calls)

also geometry nodes can join meshes / cull faces / do all sorts of interesting stuff - just enusre you are moving a empty they are targeting to update the geometry nodes, not the player / or updating constantly as it can be expensive.

**Reply by HolandonHdunto** [reply]
It's a shame for those who don't have access to mesh because it would be great for making procedural games or assembling scenes with prefabs of structures.
   It's not much, it would simply be a function that merges a group of meshes or objects, and generates a new mesh accessible by REPLACEMESH. I don't even know if I would need to merge the materials as well because you can create one material for every object in the game without using the texture atlas.
   On upbge 0.2.x.
